### ISSUE TITLE
Docker on Mac Permission Issues (External Grading)

### ISSUE THREAD

**Issue by barbehenna** [PrairieLearn/PrairieLearn#1734]
In our (myself and @eddelbuettel) external automatic grading set up we've created an unprivileged user called `ag` that runs the student code. On linux (including production) our setup correctly prevents `ag` from accessing `root` only files. However, on my mac `ag` is able to access files that have been given `root` only permissions. When this happens I'm running the mac docker locally as suggested in the PrairieLearn docs. 

This behavior appears to be by design https://docs.docker.com/docker-for-mac/osxfs/ (as stated in the access control section in the link) but is very annoying when it comes up while developing questions.

**Comment by trombonekenny** [comment]
Interesting! https://github.com/docker/for-mac/issues/2657 has a bit more back story and a few other links to consider (even if they closed it as `working-as-designed`)

**Comment by eddelbuettel** [comment]
Thanks to all for helping to get to the bottom of it.  So we're not crazy -- we're simply not operating the way Cupertino wants us to.  Ok then.   Time for my all-time-favorite comic:

![image](https://user-images.githubusercontent.com/673121/66241053-c2a33d80-e6c3-11e9-8482-e67294983bd0.png)

Source:  https://dilbert.com/strip/1995-06-24

**Comment by mwest1066** [comment]
@trombonekenny I think one solution for this might be Docker Volumes, as you were investigating earlier for Windows users?

**Comment by trombonekenny** [comment]
@mwest1066 That should be easy enough to test - that the host user permissions don't propagate to the docker volume. If so, then some sort of a script to copy things from a mounted host directory to the docker vol and adjust permissions before running?

**Comment by mwest1066** [comment]
Actually I think we already do all the copying. The problem is that the space we copy to is a mounted-in-host-filesystem (the `HOST_JOBS_DIR`) and on MacOS this doesn’t enforce permissions (argh!). We only use this space to communicate between the PL container and the grading container. If we used a Docker Volume for this instead, then my hope would be that it function as a regular linux volume and enforce permissions.

If that’s true, the only change we’d need is to tell PL to run the docker container with a volume rather than a filesystem mount, and we’d need to run the PL container with that same volume. (all of this is untested speculation)

**Comment by trombonekenny** [comment]
@barbehenna I suspect @mwest1066 is right and we could try running things as a docker volume instead of a host directory. I don't think it would be hard but since I'm not a MacOS user it would probably be best to meet in person to try things (and document what we find). If you're interested in that, let me know.

**Comment by nwalters512** [comment]
This was discussed on Slack before, going to rehash a few of my points here though. The main problem with using a Docker volume for this is that you can't mount a subdirectory of a volume into another container. However, this conflicts with how the external grader currently operates, which is that we maintain a directory per job in `HOST_JOBS_DIR` and mount a single job directory into each grading container. I believe the only way we could fix this is by serializing all grading jobs and wiping the volume in between successive jobs. However, this would remove people's ability to inspect the files after a grading job has completed, which has been invaluable to me during question development and debugging.

The above is actually what I did with https://github.com/PrairieLearn/PrairieLearn/pull/1163/files and https://github.com/PrairieLearn/PrairieGrader/pull/38/files. We could probably maintain people's ability to inspect files by optionally copying them back to somewhere on the host in between runs?

**Comment by trombonekenny** [comment]
Yeah, docker volumes by themselves aren't flexible enough to be a drop-in replacement for the HOST_JOBS_DIR. It's too bad. But from the Windows experience, and now from the Mac experience, I'm still convinced that moving to more docker-native models for this kind of interaction is better than sharing a host directory.

I assert that the convenience of having the `HOST_JOBS_DIR` local on the host to inspect for development and debugging is a minor point. The super developers like @nwalters512 will always be able to find a way to inspect the output (mount the docker volume, open another terminal on the container, worst case-- use the PL instructor/grading_job web interface.) and the other 95% of the developers never use this functionality. Worse, they suffer `docker run` command line complexities or file system access oddities because of that design choice.

I wonder if we can converge the `docker-compose` considerations with the `executor` concepts that are coming to simplify how PL launches question jobs? Localized storage in the containers and standards for intercommunication that don't care what the underlying providers or hosting are (S3 mimicker like `minio`)? With the right compose stack and config, would there still be a difference between standalone PL and docker PL?

**Comment by reteps** [comment]
I have been able to get external grading to work on my Mac, using both a native method and docker containers. Thus, I am considering this completed.
